monai/transforms/io/array.py

```diff
@@ -57,109 +68,133 @@ def switch_endianness(data, new="<"):
-                a reader object with the `*args` and `**kwargs` parameters, supported reader name: "NibabelReader",
-                "PILReader", "ITKReader", "NumpyReader".
+            reader: reader to load image file and meta data
+
+                - if `reader` is None, a default set of `SUPPORTED_READERS` will be used.
+                - if `reader` is a string, the corresponding item in `SUPPORTED_READERS` will be used,
+                  and a reader instance will be constructed with the `*args` and `**kwargs` parameters.
+                  the supported reader names are: "nibabelreader", "pilreader", "itkreader", "numpyreader".
+                - if `reader` is a reader class/instance, it will be registered to this loader accordingly.
+
             image_only: if True return only the image volume, otherwise return image data array and header dict.
             dtype: if not None convert the loaded image to this data type.
             args: additional parameters for reader if providing a reader name.
             kwargs: additional parameters for reader if providing a reader name.
 
         Note:
-            The transform returns image data array if `image_only` is True,
-            or a tuple of two elements containing the data array, and the meta data in a dict format otherwise.
+
+            - The transform returns an image data array if `image_only` is True,
+              or a tuple of two elements containing the data array, and the meta data in a dictionary format otherwise.
+            - If `reader` is specified, the loader will attempt to use the specified readers and the default supported
+              readers. This might introduce overheads when handling the exceptions of trying the incompatible loaders.
+              In this case, it is therefore recommended to set the most appropriate reader as
+              the last item of the `reader` parameter.
 
         """
-        # set predefined readers as default
-        self.readers: List[ImageReader] = [ITKReader(), NumpyReader(), PILReader(), NibabelReader()]
-        if reader is not None:
-            if isinstance(reader, str):
-                supported_readers = {
-                    "nibabelreader": NibabelReader,
-                    "pilreader": PILReader,
-                    "itkreader": ITKReader,
-                    "numpyreader": NumpyReader,
-                }
-                the_reader = look_up_option(reader.lower(), supported_readers)
-                self.register(the_reader(*args, **kwargs))
-            else:
-                self.register(reader)
 
+        self.auto_select = reader is None
         self.image_only = image_only
         self.dtype = dtype
 
-    def register(self, reader: ImageReader) -> List[ImageReader]:
+        self.readers: List[ImageReader] = []
+        for r in SUPPORTED_READERS:  # set predefined readers as default
+            try:
+                self.register(SUPPORTED_READERS[r](*args, **kwargs))
```

Comment:
now we support adding new readers to `monai.transforms.SUPPORTED_READERS` at the package level (instead of registering at each reader instance), so I think it'll be good to be flexible here

Reply:
yes, I think the exceptions are handled here, or could you please construct a testing case to show the bug?

Reply:
Hi @Nic-Ma I'll mark this as resolved for now, please open new tickets for further issues, I can look into them.